packages/cozy-sharing/src/components/ShareRecipientsInput.jsx

```diff
@@ -1,73 +1,147 @@
-import get from 'lodash/get'
 import PropTypes from 'prop-types'
-import React, { useState, useEffect } from 'react'
+import React, { useEffect, useMemo, useState } from 'react'
+
+import { isQueryLoading, useQueryAll } from 'cozy-client'
+import flag from 'cozy-flags'
 
 import ShareAutosuggest from './ShareAutosuggest'
-import { contactsResponseType, contactGroupsResponseType } from '../propTypes'
+import { getContactsFromGroupId } from '../helpers/contacts'
+import {
+  buildReachableContactsQuery,
+  buildContactGroupsByIdsQuery,
+  buildUnreachableContactsFromGroupsQuery
+} from '../queries/queries'
 
+/**
+ * ShareRecipientsInput is responsible for fetching contacts and groups.
+ * We retrieve all the contacts that are reachable and the groups they belong to.
+ * In order to display the total number of members in each group, we also retrieve the contacts that are not reachable.
+ *
+ * We only show the loading state when the user has focus on the input.
+ */
 const ShareRecipientsInput = ({
-  contacts,
-  contactGroups,
   recipients,
   placeholder,
   onPick,
   onRemove
 }) => {
-  const [loading, setLoading] = useState(false)
+  const [isLoading, setLoading] = useState(false)
+  const reachableContactsQuery = buildReachableContactsQuery()
+  const reachableContactsResult = useQueryAll(
+    reachableContactsQuery.definition,
+    reachableContactsQuery.options
+  )
+
+  const groupIdsFromReachableContacts = useMemo(() => {
+    return [
+      ...new Set(
+        (reachableContactsResult.data || []).flatMap(contact => {
+          return (contact.relationships?.groups?.data || []).map(
+            group => group._id
+          )
+        })
+      )
+    ]
+  }, [reachableContactsResult])
+
+  const contactGroupsByIdsQuery = buildContactGroupsByIdsQuery(
+    groupIdsFromReachableContacts
+  )
+  const contactGroupsByIdsResult = useQueryAll(
+    contactGroupsByIdsQuery.definition,
+    contactGroupsByIdsQuery.options
+  )
+
+  const unreachableContactsFromGroupsQuery =
+    buildUnreachableContactsFromGroupsQuery(groupIdsFromReachableContacts)
+
+  const unreachableContactsFromGroupsResult = useQueryAll(
+    unreachableContactsFromGroupsQuery.definition,
+    {
+      ...unreachableContactsFromGroupsQuery.options,
+      enabled:
+        groupIdsFromReachableContacts.length > 0 &&
+        flag('sharing.show-recipient-groups')
+    }
+  )
 
   useEffect(() => {
     if (
-      loading &&
-      !contacts.hasMore &&
-      contacts.fetchStatus === 'loaded' &&
-      !contactGroups.hasMore &&
-      contactGroups.fetchStatus === 'loaded'
+      !reachableContactsResult.hasMore &&
+      reachableContactsResult.fetchStatus === 'loaded' &&
+      !contactGroupsByIdsResult.hasMore &&
+      contactGroupsByIdsResult.fetchStatus === 'loaded' &&
+      (flag('sharing.show-recipient-groups') !== true ||
+        (!unreachableContactsFromGroupsResult.hasMore &&
+          unreachableContactsFromGroupsResult.fetchStatus === 'loaded'))
```

Comment:
nit: this `if` starts to be hard to read, maybe we can split it into multiple named variables? (except if the performances are important but seems like it would have only few impact)

Reply:
also can't we use `isQueryLoading` here like in the following block?

Reply:
I found a better way to make it readable. I've separated the responsibilities. This component will only be responsible for knowing if there is data loading. And the component below will be responsible for displaying the loading status only if the user has focused the input. This allows me to remove this useEffect from the component 🙂